Add support for duplicate attributes to the HTML fast parser

cdumez opened the pull request:
#### 55036f71e77686c94215a6963d5e36576faa59ee
<pre>
Add support for duplicate attributes to the HTML fast parser
<a href="https://bugs.webkit.org/show_bug.cgi?id=255618">https://bugs.webkit.org/show_bug.cgi?id=255618</a>

Reviewed by Darin Adler.

The HTML fast parser already had logic to detect duplicate attributes, so it
was already doing most the work. However, when finding a duplicate, it would
just abort and let the full HTML parser deal with the input string. It is not
much more work to properly support duplicate attributes and it doesn&apos;t slow
down the parser on Speedometer. As a result, I think we should add the
functionality.

Note that I am still using a Vector to store the attribute names. Switching to
a HashSet regresses Speedometer. I am however keeping the Vector sorted to
speed up detection of duplicates.

* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::insertInUniquedSortedVector):
(WebCore::HTMLFastPathParser::parseAttributes):

Canonical link: <a href="https://commits.webkit.org/263136@main">https://commits.webkit.org/263136@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59ae5378471e598169c15bbe6c712aff87bbeb4d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3622 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3663 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3812 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5048 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3910 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3596 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3763 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3711 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3143 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3666 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3904 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3263 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4926 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1410 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3235 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/3337 "142 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3212 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3294 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4688 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3675 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2973 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3209 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3235 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/906 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3241 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3488 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -123,6 +123,15 @@ template<typename CharacterType> static inline bool isCharAfterUnquotedAttribute
 
+template<typename T> static bool insertInUniquedSortedVector(Vector<T>& vector, const T& value)
+{
+    auto it = std::lower_bound(vector.begin(), vector.end(), value);
```
I am guessing that using binary search to insert into a kept-sorted vector is the best way to go, but I can’t help wondering if there is a better data structure and algorithm for checking for duplicates. For example, when the vector is very short, a linear search will be faster than a binary search, but of course then it won’t be sorted for when the vector gets longer.

cdumez replied:
I tried both linear search (unsorted vector) and binary search with a sorted vector. They were both perf neutral on Speedometer (unlike the HashSet which was a regression). I went with the sorted vector / binary search because it should be faster with larger vectors and there was no cost on Speedometer.

cdumez committed:
Add support for duplicate attributes to the HTML fast parser
https://bugs.webkit.org/show_bug.cgi?id=255618

Reviewed by Darin Adler.

The HTML fast parser already had logic to detect duplicate attributes, so it
was already doing most the work. However, when finding a duplicate, it would
just abort and let the full HTML parser deal with the input string. It is not
much more work to properly support duplicate attributes and it doesn't slow
down the parser on Speedometer. As a result, I think we should add the
functionality.

Note that I am still using a Vector to store the attribute names. Switching to
a HashSet regresses Speedometer. I am however keeping the Vector sorted to
speed up detection of duplicates.

* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::insertInUniquedSortedVector):
(WebCore::HTMLFastPathParser::parseAttributes):

Canonical link: https://commits.webkit.org/263136@main
